### DISCUSSION TITLE
Executing .wasm policies using OPA server

### DISCUSSION THREAD

**Discussion by dushyant7917** [open-policy-agent/community#293] [Wasm] [answered]
OPA server provides out of the box facility to execute Rego policies. OPA cli provides support to compile Rego policies into .wasm files. Can we use OPA server to execute the .wasm policies? If yes, is there any tutorial documentation for this?

**Comment by srenatus** [comment, marked as answer]
If you use the default images (non-static), then `opa run` is able to work with the bundles built using `opa build -t wasm -e some/entrypoint ...`. Those bundles contain a `policy.wasm`; and there's an SDK in OPA that'll use wasmtime to evaluate them. It should be fairly transparent, but the wasm modules always stay inside bundles. OPA does not accept plain .wasm files.

**Reply by dushyant7917** [reply]
What I want to ask is -> right now if we use OPA server by using any of the examples mentioned [here](https://www.openpolicyagent.org/docs/latest/rest-api/#query-api), OPA server is executing standard Rego policies? How can we configure OPA server to use WASM policies instead of Rego for any REST API call?

**Reply by srenatus** [reply]
There's some mechanism in the OPA server that'll make it work:

1. Build wasm bundle, run server:
```interactive
$ cat foo.rego 
package foo
import future.keywords.if

bar := "baz" if input == "bar"
$ opa build -t wasm -e foo/bar foo.rego 
$ opa run --log-level debug --server -a 127.0.0.1:8181 -b bundle.tar.gz 
{"addrs":["127.0.0.1:8181"],"diagnostic-addrs":[],"level":"info","msg":"Initializing server.","time":"2022-10-27T20:07:16+02:00"}
{"level":"debug","msg":"maxprocs: Leaving GOMAXPROCS=16: CPU quota undefined","time":"2022-10-27T20:07:16+02:00"}
{"level":"debug","msg":"Server initialized.","time":"2022-10-27T20:07:16+02:00"}
```
2. In another terminal, use the v1 Data API:
```interactive
$ curl -v "http://127.0.0.1:8181/v1/data/foo/bar?metrics&pretty" -d '{"input": "bar"}'       
*   Trying 127.0.0.1:8181...
* Connected to 127.0.0.1 (127.0.0.1) port 8181 (#0)
> POST /v1/data/foo/bar?metrics&pretty HTTP/1.1
> Host: 127.0.0.1:8181
> User-Agent: curl/7.79.1
> Accept: */*
> Content-Length: 16
> Content-Type: application/x-www-form-urlencoded
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: application/json
< Date: Thu, 27 Oct 2022 18:08:54 GMT
< Content-Length: 452
< 
{
  "metrics": {
    "counter_server_query_cache_hit": 1,
    "timer_rego_external_resolve_ns": 160582,
    "timer_rego_input_parse_ns": 12883,
    "timer_rego_query_eval_ns": 199741,
    "timer_server_handler_ns": 250208,
    "timer_wasm_pool_acquire_ns": 1153,
    "timer_wasm_pool_release_ns": 560,
    "timer_wasm_vm_eval_call_ns": 70233,
    "timer_wasm_vm_eval_ns": 77522,
    "timer_wasm_vm_eval_prepare_input_ns": 4353
  },
  "result": "baz"
}
```

☝️ As you can see from the metrics, the internal wasm SDK was used for evaluting the wasm bundle to reply to that ordinary v1 Data API request.

**Reply by dushyant7917** [reply]
Thanks, this is what I was looking for!
But why the server is evaluating the .rego policy as well from bundle.tar.gz along with executing .wasm policy? From the metrics I can see that rego evaluation is also taking place!
Can we disable rego execution completely and just use wasm execution?

**Reply by dushyant7917** [reply]
> Why the server is evaluating the .rego policy as well from bundle.tar.gz along with executing .wasm policy? From the metrics I can see that rego evaluation is also taking place! Can we disable rego execution completely and just use wasm execution?

@srenatus ^^ bump

**Reply by srenatus** [reply]
Sorry I've forgotten about this one, it seems. 😳

You might have to define more entrypoints.

**Reply by dushyant7917** [reply]
> You might have to define more entrypoints.

Didn't understood this point?

**Reply by dushyant7917** [reply]
@srenatus Bump ^^

**Reply by srenatus** [reply]
Nevermind. I've tried passing `-e foo` in addition to `-e foo/bar` but it doesn't change the outcome.

> Can we disable rego execution completely and just use wasm execution?

No you cannot. But don't worry about rego execution -- topdown is the entrypoint that the server is using, and there's no way around that. But as you can see from the trace, it'll _immediately_ jump into the Wasm module:

```interactive
$ echo '"bar"' | opa eval -I -fpretty -b bundle.tar.gz --explain=full data.foo.bar
query:1     Enter data.foo.bar = _
query:1     | Eval data.foo.bar = _
query:1     | Wasm data.foo
query:1     | Exit data.foo.bar = _
query:1     Redo data.foo.bar = _
query:1     | Redo data.foo.bar = _
"baz"
```
